server/route.go

```diff
@@ -62,12 +61,196 @@ type connectInfo struct {
+	s.rqsMu.Unlock()
+}
+
+// This is for when we receive a directed message for a queue subscriber
+// that has gone away. We reroute like a new message but scope to only
+// the queue subscribers that it was originally intended for. We will
+// prefer local clients, but will bounce to another route if needed.
+func (c *client) reRouteQMsg(r *SublistResult, msgh, msg, group []byte) {
+	c.Debugf("Attempting redelivery of message for absent queue subscriber on group '%q'", group)
+
+	// We only care about qsubs here. Data structure not setup for optimized
+	// lookup for our specific group however.
+
+	var qsubs []*subscription
+	for _, qs := range r.qsubs {
+		if len(qs) != 0 && bytes.Equal(group, qs[0].queue) {
+			qsubs = qs
+			break
+		}
+	}
+
+	// If no match return.
+	if qsubs == nil {
+		c.Debugf("Redelivery failed, no queue subscribers for message on group '%q'", group)
+		return
+	}
+
+	// We have a matched group of queue subscribers.
+	// We prefer a local subscriber since that was the original target.
+
+	// Spin prand if needed.
+	if c.in.prand == nil {
+		c.in.prand = rand.New(rand.NewSource(time.Now().UnixNano()))
+	}
+
+	// Hold onto a remote if we come across it to utilize in case no locals exist.
+	var rsub *subscription
+
+	startIndex := c.in.prand.Intn(len(qsubs))
+	for i := 0; i < len(qsubs); i++ {
+		index := (startIndex + i) % len(qsubs)
+		sub := qsubs[index]
+		if sub == nil {
+			continue
+		}
+		if rsub == nil && bytes.HasPrefix(sub.sid, []byte(QRSID)) {
+			rsub = sub
+			continue
+		}
+		mh := c.msgHeader(msgh[:], sub)
+		if c.deliverMsg(sub, mh, msg) {
+			c.Debugf("Redelivery succeeded for message on group '%q'", group)
+			return
+		}
+	}
+	// If we are here we failed to find a local, see if we snapshotted a
+	// remote sub, and if so deliver to that.
+	if rsub != nil {
+		mh := c.msgHeader(msgh[:], rsub)
+		if c.deliverMsg(rsub, mh, msg) {
+			c.Debugf("Re-routing message on group '%q' to remote server", group)
+			return
+		}
+	}
+	c.Debugf("Redelivery failed, no queue subscribers for message on group '%q'", group)
+}
+
+// processRoutedMsg processes messages inbound from a route.
+func (c *client) processRoutedMsg(r *SublistResult, msg []byte) {
+	// Snapshot server.
+	srv := c.srv
+
+	msgh := c.prepMsgHeader()
+	si := len(msgh)
+
+	// If we have a queue subscription, deliver direct
+	// since they are sent direct via L2 semantics over routes.
+	// If the match is a queue subscription, we will return from
+	// here regardless if we find a sub.
+	isq, sub, err := srv.routeSidQueueSubscriber(c.pa.sid)
+	if isq {
+		if err != nil {
+			// We got an invalid QRSID, so stop here
+			c.Errorf("Unable to deliver routed queue message: %v", err)
+			return
+		}
+		didDeliver := false
+		if sub != nil {
+			mh := c.msgHeader(msgh[:si], sub)
+			didDeliver = c.deliverMsg(sub, mh, msg)
+		}
+		if !didDeliver && c.srv != nil {
+			group := c.srv.lookupRemoteQGroup(string(c.pa.sid))
+			c.reRouteQMsg(r, msgh, msg, group)
```

Comment:
I think that we should still account for group being not found and in this case simply return.

Reply:
That is handled via a DebugF on route.go:159